### ISSUE TITLE
Invalid reads from DOS Exe Header?

### ISSUE THREAD

**Issue by LowLevelMahn** [uxmal/reko#788]
everything after: e_ovno is not part of the normal classic header, and you got NE an PE header loaders - so what are the members, reads for?

https://github.com/uxmal/reko/blob/3a9b471f92f830dc86148c632e59507566553a27/src/ImageLoaders/MzExe/ExeImageLoader.cs#L164-L176

these fields are not available in classic dos executables (no NE) or filled with anything the linker wants

**Layout:**
```
DOS-Header
...
Relocation-table
...
Image
```

for example:

**single.asm**
```
; build with: 
;  nasm.exe (https://www.nasm.us/pub/nasm/releasebuilds/2.14.02/)
;
;  nasm.exe -f obj -o single.obj single.asm

BITS 16

segment seg000 align=16

text: db 'Hello World!',0ah,0dh,'$'

segment seg001 align=16

..start:
mov ax,seg000
mov ds,ax

push ax
pop ax

call far print

mov ax,0x4c00
int 0x21

segment seg002 align=16

print:
mov dx,text
mov ah,9
int 0x21
retf

segment seg003 stack
resb 256
```

**build with these linkers:**
```
wlink.exe: Open Watcom Linker Version 2.0 beta Sep 13 2019 01:44:55 (64-bit)
link.exe: Microsoft (R) Segmented Executable Linker Version 5.60.339 Dec 5 1994
optlink.exe: OPTLINK (R) for Win32 Release 8.00.17 (from the dmd package: dmd.2.088.0.windows)
ulink.exe: UniLink v1.11 [beta] (build 11.27) from ftp://ftp.styx.cabel.net/pub/UniLink/
```

**data between header and relocation table by linker:**
```
optlink.exe: 00 00
link.exe: 01 00
wlink.exe: 00 00 00 00
ulink.exe:
00000000  55 6E 69 4C 69 6E 6B 00 00 00 00 00 00 00 00 00  UniLink.........
00000010  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
00000020  00 00 00 00                                                             ....
```

above single.asm build with all the linkers:
http://s000.tinyupload.com/?file_id=57475334987014459232

so you are reading parts of the image, alignment bytes or linker data

**Comment by uxmal** [comment]
Looks like you're right. The "old" MS-DOS header is 0x1C bytes long, ending with `e_ovno`. The `e_lfaNew` field, at offset 0x3C, is only valid if the `e_lfaRelocations` field contains a value of 0x0040 or larger. The ExeImageLoader needs to be changed to only read up to `e_ovno` bytes. Then, only if the `e_lfaRelocations` is larger than or equal to 0x40, should it attempt to load `NE` or `PE` executables. Otherwise, it should fall back to real-mode MS-DOS loading code.

**Comment by uxmal** [comment]
Looking at some PE executables I have in my "zoo", it seems Windows is content to load PE executables whose MS-DOS headers have `e_lfaRelocations` set to 0.  The modified logic will thus be:
1. Read only 0x1A bytes of the header are read, not 0x40.
2. If the executable file is >= 0x40 bytes then it is safe to read the `e_lfaNew` field at file offset 0x3C. 
3. The read `e_lfaNew` value is used to locate the `NE` or `PE` signature, as before. If the signatures are present, dispatch to the appropriate NE/PE loader
4. Otherwise, load real-mode MS-DOS image.

**Comment by LowLevelMahn** [comment]
>The modified logic will thus be...

looks good

> PE executables whose MS-DOS headers have e_lfaRelocations set to 0.
i don't know for sure but isn't it typical to have 0 relocations in the DOS classic part of the header in PE files?

**Comment by uxmal** [comment]
Indeed. I was reading documentation that seemed to imply that PE/NE binaries would have `e_lfarelocations` set to 0x40 or more, when clearly that's not the case. The new logic has been implemented in commit 5c7af0f.
